uamqp/aio/_cbs_async.py

```diff
@@ -0,0 +1,227 @@
+#-------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+#-------------------------------------------------------------------------
+
+import logging
+import asyncio
+from datetime import datetime
+
+from ._management_link_async import ManagementLink
+from ..utils import utc_now, utc_from_timestamp
+from ..message import Message, Properties
+from ..error import (
+    AuthenticationException,
+    TokenAuthFailure,
+    TokenExpired
+)
+from ..constants import (
+    CbsState,
+    CbsAuthState,
+    CBS_PUT_TOKEN,
+    CBS_EXPIRATION,
+    CBS_NAME,
+    CBS_TYPE,
+    CBS_OPERATION,
+    ManagementExecuteOperationResult,
+    ManagementOpenResult,
+    DEFAULT_AUTH_TIMEOUT
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def is_coroutine(get_token):
+    try:
+        if asyncio.iscoroutinefunction(get_token.func):
+            return True
+    except AttributeError:
+        if asyncio.iscoroutinefunction(get_token):
+            return True
+    raise ValueError("get_token must be a coroutine function")
+
+
+class CBSAuthenticator(object):
+    def __init__(
+        self,
+        session,
+        auth,
+        **kwargs
+    ):
+        self._session = session
+        self._connection = self._session._connection
+        self._mgmt_link = self._session.create_request_response_link_pair(
+            endpoint='$cbs',
+            on_amqp_management_open_complete=self._on_amqp_management_open_complete,
+            on_amqp_management_error=self._on_amqp_management_error,
+            status_code_field=b'status-code',
+            status_description_field=b'status-description'
+        )  # type: ManagementLink
+
+        if not auth.get_token or not is_coroutine(auth.get_token):
+            raise ValueError("get_token must be a coroutine object.")
+
+        self._auth = auth
+        self._encoding = 'UTF-8'
+        self._auth_timeout = kwargs.pop('auth_timeout', DEFAULT_AUTH_TIMEOUT)
+        self._token_put_time = None
+        self._expires_in = None
+        self._expires_on = None
+        self._token = None
+
+        self._refresh_window = None
+
+        self._token_status_code = None
+        self._token_status_description = None
+
+        self.state = CbsState.CLOSED
+        self.auth_state = CbsAuthState.Idle
+
+    async def _put_token(self, token, token_type, audience, expires_on=None):
+        # type: (str, str, str, datetime) -> None
+        message = Message(
+            value=token,
+            properties=Properties(message_id=self._mgmt_link.next_message_id),
```

Comment:
Wasn't the management link PR updating the message ID on the outgoing mgmt request?

Reply:
I can't remember which PR I wrote, but I have updated the code to duplicate what we have in sync module -- update next_message_id to be +1 after sending out the request, and it works